lib/src/util/side_effect_visitor.dart

```diff
@@ -0,0 +1,546 @@
+import 'package:analyzer/analyzer.dart';
+
+class SideEffectVisitor extends RecursiveAstVisitor {
+  var _hasSideEffect = false;
+
+  bool get hasSideEffect => _hasSideEffect;
+  @override
+  visitAdjacentStrings(AdjacentStrings node) {
+    super.visitAdjacentStrings(node);
+  }
+
+  @override
+  visitAnnotation(Annotation node) {
+    super.visitAnnotation(node);
+  }
+
+  @override
+  visitArgumentList(ArgumentList node) {
+    super.visitArgumentList(node);
+  }
+
+  @override
+  visitAsExpression(AsExpression node) {
+    super.visitAsExpression(node);
+  }
+
+  @override
+  visitAssertStatement(AssertStatement node) {
+    super.visitAssertStatement(node);
+  }
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitAwaitExpression(AwaitExpression node) {
+    super.visitAwaitExpression(node);
+  }
+
+  @override
+  visitBinaryExpression(BinaryExpression node) {
+    super.visitBinaryExpression(node);
+  }
+
+  @override
+  visitBlock(Block node) {
+    super.visitBlock(node);
+  }
+
+  @override
+  visitBlockFunctionBody(BlockFunctionBody node) {
+    super.visitBlockFunctionBody(node);
+  }
+
+  @override
+  visitBooleanLiteral(BooleanLiteral node) {
+    super.visitBooleanLiteral(node);
+  }
+
+  @override
+  visitBreakStatement(BreakStatement node) {
+    super.visitBreakStatement(node);
+  }
+
+  @override
+  visitCascadeExpression(CascadeExpression node) {
+    super.visitCascadeExpression(node);
+  }
+
+  @override
+  visitCatchClause(CatchClause node) {
+    super.visitCatchClause(node);
+  }
+
+  @override
+  visitClassDeclaration(ClassDeclaration node) {
+    super.visitClassDeclaration(node);
+  }
+
+  @override
+  visitClassTypeAlias(ClassTypeAlias node) {
+    super.visitClassTypeAlias(node);
+  }
+
+  @override
+  visitComment(Comment node) {
+    super.visitComment(node);
+  }
+
+  @override
+  visitCommentReference(CommentReference node) {
+    super.visitCommentReference(node);
+  }
+
+  @override
+  visitCompilationUnit(CompilationUnit node) {
+    super.visitCompilationUnit(node);
+  }
+
+  @override
+  visitConditionalExpression(ConditionalExpression node) {
+    super.visitConditionalExpression(node);
+  }
+
+  @override
+  visitConfiguration(Configuration node) {
+    super.visitConfiguration(node);
+  }
+
+  @override
+  visitConstructorDeclaration(ConstructorDeclaration node) {
+    super.visitConstructorDeclaration(node);
+  }
+
+  @override
+  visitConstructorFieldInitializer(ConstructorFieldInitializer node) {
+    super.visitConstructorFieldInitializer(node);
```

Comment:
I don't think you should ever visit one of these, but a constructor field initializer *does* have the side effect of assigning a value to a field.